src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+        let mut mfp_eval = MfpEval {
+            evaluator,
+            input_arity,
+            expressions: vec![],
+        };
+        for expr in expressions {
+            let result = mfp_eval.eval_expr(expr);
+            mfp_eval.expressions.push(result);
+        }
+        mfp_eval
+    }
+}
+
+impl<'a, E: Interpreter + ?Sized> Interpreter for MfpEval<'a, E> {
+    type Summary = E::Summary;
+
+    fn eval_column(&self, id: usize) -> Self::Summary {
+        if id < self.input_arity {
+            self.evaluator.eval_column(id)
+        } else {
+            self.expressions[id - self.input_arity].clone()
+        }
+    }
+
+    fn eval_literal(
+        &self,
+        result: &Result<Row, EvalError>,
+        col_type: &ColumnType,
+    ) -> Self::Summary {
+        self.evaluator.eval_literal(result, col_type)
+    }
+
+    fn eval_unmaterializable(&self, func: &UnmaterializableFunc) -> Self::Summary {
+        self.evaluator.eval_unmaterializable(func)
+    }
+
+    fn eval_unary(&self, func: &UnaryFunc, expr: Self::Summary) -> Self::Summary {
+        self.evaluator.eval_unary(func, expr)
+    }
+
+    fn eval_binary(
+        &self,
+        func: &BinaryFunc,
+        left: Self::Summary,
+        right: Self::Summary,
+    ) -> Self::Summary {
+        self.evaluator.eval_binary(func, left, right)
+    }
+
+    fn eval_variadic(&self, func: &VariadicFunc, exprs: Vec<Self::Summary>) -> Self::Summary {
+        self.evaluator.eval_variadic(func, exprs)
+    }
+
+    fn eval_if(
+        &self,
+        cond: Self::Summary,
+        then: Self::Summary,
+        els: Self::Summary,
+    ) -> Self::Summary {
+        self.evaluator.eval_if(cond, then, els)
+    }
+}
+
+/// For a given column and expression, does the range of the column influence the range of the
+/// expression? (ie. if the expression is a filter, could we tell that the filter
+/// will never return True?)
+#[derive(Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Debug)]
+pub enum Pushdownable {
+    No,
+    Maybe,
+    Yes,
+}
+
+impl Pushdownable {
+    fn apply_fn(mut columns: Vec<Self>, is_monotonic: Monotonic) -> Vec<Self> {
+        match is_monotonic {
+            Monotonic::Yes => {
+                // Still correlated, nothing to do.
+            }
+            Monotonic::No => {
+                // A function that we know we can't see through. No reason to expect the range
+                // of the columns to affect the range of the expression.
+                for col in &mut columns {
+                    *col = Pushdownable::No;
+                }
+            }
+            Monotonic::Maybe => {
+                // A function that we may or may not be able to see through.
+                for col in &mut columns {
+                    *col = Pushdownable::Maybe.min(*col);
+                }
+            }
+        }
+        columns
+    }
+
+    fn merge(mut columns: Vec<Self>, other: Vec<Self>) -> Vec<Self> {
```

Comment:
I may be the only one with this problem, but this name feels ambiguous to me---is `merge` a least upper bound or a greatest lower bound?

I'm also confused by the `Vec::resize` call... when would we have information about different numbers of columns that we could sensibly combine?

Reply:
Least upper, unless I'm thinking about this totally wrong! `union` seems like a fine and more specific name; will switch to that and you can tell me if it's worse. :)

Re: `resize`: this is an artifact of:
- `Trace` doesn't know the number of columns, so we can't size the vector correctly up front.
- We don't need to distinguish between columns that don't exist or columns that aren't pushdownable. (In `EXPLAIN` output, we only need to list the columns that _are_ pushdownable.)
So effectively I'm treating this `Vec<Pushdownable` as a `Map<usize, Pushdownable>`. Would it make sense to just using a `Map`? A bit less memory-efficient, but this isn't on a hot path, so we can afford the clarity.

Reply:
Got it---would be good to put a comment to explain the invariant on what's in the `Vec`, if only to remind me when I see this again. 😁 

I've got no prior on how many columns we can expect in a query, but I suspect things would have to get quite big---in the 100s of columns---for a `Map` to be better performance-wise. But developer understanding-wise, it might be better to treat it as a `Vec<Option<Pushdownable>>` to make it clear whether we don't know anything for a column vs. knowing affirmatively something about it.

Reply:
Got it! I've added a wrapper type in a new commit to encapsulate this behaviour and to have a more obvious place to hang documentation off of.